File: LeetCode/3Sum.cpp
#include <iostream>
#include <vector>
#include <map>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<vector<int> > threeSum(vector<int> &num) {
        vector<vector<int> > ret;
        if (num.size() < 3) return ret;

        sort(num.begin(), num.end());

        for (int i = 0; i < num.size() - 2; ++i)
        {
            if (i != 0 && num[i] == num[i-1]) continue;
            int start = i + 1, end = num.size() - 1;
            int target = 0 - num[i];
            while (start < end){
                if (start != i+1 && num[start] == num[start-1]){
                    start++;
                    continue;
                }
                if (num[start] + num[end] > target){
                    end--;
                }else if (num[start] + num[end] < target){
                    start++;
                }else{
                    vector<int>  ans;
                    ans.push_back(num[i]);
                    ans.push_back(num[start]);
                    ans.push_back(num[end]);
                    ret.push_back(ans);
                    end--;
                    start++;
                }
            }
        }

        return ret;
    }
};

int main(int argc, char const *argv[])
{
    Solution sol;
    vector<int> num;
    num.push_back(0);
    num.push_back(0);
    num.push_back(0);
    vector<vector<int> > ans = sol.threeSum(num);
    for (int i = 0; i < ans.size(); ++i)
    {
        for (int j = 0 ; j < ans[i].size(); j++){
            cout << ans[i][j] << "\t";
        }
        cout << endl;
    }
    return 0;
}

File: LeetCode/3Sum.py
class Solution:
	# @return a list of lists of length 3, [[val1,val2,val3]]
	def threeSum(self, num):
		l = len(num)
		if l == 0:
			return []
		num.sort()
		ans = set()
		for i in range(0, l-2):
			j = i + 1
			k = l - 1	
			while k > j:
				sum3 = num[i] + num[j] + num[k]
	
				if (sum3 == 0):
					ans.add((num[i], num[j], num[k])) 
					j += 1
					k -= 1
				elif (sum3 > 0):
					k -= 1
				elif (sum3 < 0):
					j += 1

		return [list(i) for i in ans]


if __name__ == '__main__':
	num = [-1, 0, 1, 2, -1, -4]
	s = Solution()
	print s.threeSum(num)
